include/ensmallen_bits/eve/eve_impl.hpp

```diff
@@ -0,0 +1,169 @@
+  size_t currentFunction = 0;
+  double overallObjective = 0;
+  double lastOverallObjective = DBL_MAX;
+
+  double objective = 0;
+  double lastObjective = 0;
+  double dt = 1;
+
+  // The exponential moving average of gradient values.
+  arma::mat m = arma::zeros<arma::mat>(iterate.n_rows, iterate.n_cols);
+
+  // The exponential moving average of squared gradient values.
+  arma::mat v = arma::zeros<arma::mat>(iterate.n_rows, iterate.n_cols);
+
+  // Now iterate!
+  arma::mat gradient(iterate.n_rows, iterate.n_cols);
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+      std::numeric_limits<size_t>::max() : maxIterations;
+  for (size_t i = 0; i < actualMaxIterations; /* incrementing done manually */)
+  {
+    // Is this iteration the start of a sequence?
+    if ((currentFunction % numFunctions) == 0 && i > 0)
+    {
+      // Output current objective function.
+      Info << "Eve: iteration " << i << ", objective " << overallObjective
+          << "." << std::endl;
+
+      if (std::isnan(overallObjective) || std::isinf(overallObjective))
+      {
+        Warn << "Eve: converged to " << overallObjective << "; terminating"
+            << " with failure.  Try a smaller step size?" << std::endl;
+        return overallObjective;
+      }
+
+      if (std::abs(lastOverallObjective - overallObjective) < tolerance)
+      {
+        Info << "Eve: minimized within tolerance " << tolerance << "; "
+            << "terminating optimization." << std::endl;
+        return overallObjective;
+      }
+
+      // Reset the counter variables.
+      lastOverallObjective = overallObjective;
+      overallObjective = 0;
+      currentFunction = 0;
+
+      if (shuffle) // Determine order of visitation.
+        f.Shuffle();
+    }
+
+    // Find the effective batch size; we have to take the minimum of three
+    // things:
+    // - the batch size can't be larger than the user-specified batch size;
+    // - the batch size can't be larger than the number of iterations left
+    //       before actualMaxIterations is hit;
+    // - the batch size can't be larger than the number of functions left.
+    const size_t effectiveBatchSize = std::min(
+        std::min(batchSize, actualMaxIterations - i),
+        numFunctions - currentFunction);
+
+    // Technically we are computing the objective before we take the step, but
+    // for many FunctionTypes it may be much quicker to do it like this.
+    objective = f.EvaluateWithGradient(iterate, currentFunction,
+        gradient, effectiveBatchSize);
+    overallObjective += objective;
+
+    m *= beta1;
+    m += (1 - beta1) * gradient;
+
+    v *= beta2;
+    v += (1 - beta2) * (gradient % gradient);
+
+    const double biasCorrection1 = 1.0 - std::pow(beta1, (double) (i + 1));
+    const double biasCorrection2 = 1.0 - std::pow(beta2, (double) (i + 1));
+
+    if (i > 0)
+    {
+      const double d = std::abs(objective - lastObjective) /
+          std::min(objective, lastObjective);
```

Comment:
I wonder if this is the source of the nan in the tests.  Maybe it needs some correction in the denominator?

Reply:
Right, ``std::min(objective, lastObjective)`` could be zero.